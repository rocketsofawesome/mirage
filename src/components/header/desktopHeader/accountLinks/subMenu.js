import React from 'react'
import styled, { css, keyframes } from 'styled-components'
import { CSSTransitionGroup } from 'react-transition-group'
import { Link } from 'SRC'
import PropTypes from 'prop-types'

const transition = {
  enter: 'enter',
  leave: 'leave'
}

const slideDown = keyframes`
  0% {
    transform: translateY(-60px);
  }
  100% {
    transform: translateY(-1px);
  }
`
const position = css`
position: absolute;
transform: translateY(-1px);
top: 100%;
right: -1px;
z-index: 10;
.subMenu {
  position: absolute;
  top: 100%;
  right: 0;
  flex-wrap: wrap;
  height: auto;
  > * {
    width: 100%;
  }
}
`

export class BaseSubMenu extends React.Component {
  render () {
    const { className, childCount, open, signOut, isSubscriptionMember, renderLink, pathname } = this.props
    return (
      <CSSTransitionGroup
        transitionName={transition}
        transitionEnterTimeout={250}
        transitionLeaveTimeout={250}
        className={className}>
        {open &&
          <nav className='subMenu'>
            {isSubscriptionMember &&
              <div>
                <Link
                  uppercase
                  underline={false}
                  target='/style-file'
                  renderLink={renderLink}>
                  Style File
                </Link>
                <Link
                  uppercase
                  underline={false}
                  target='/deliveries'
                  renderLink={renderLink}>
                  Manage Deliveries
                </Link>
                <Link
                  uppercase
                  underline={false}
                  target='/reverse'
                  renderLink={renderLink}
                  className={pathname === '/reverse' && 'highlighted'}
                >
                  Rockets Reverse
                </Link>
              </div>
            }
            <Link
              uppercase
              underline={false}
              target='/orders'
              renderLink={renderLink}>
              Order History
            </Link>
            <Link
              uppercase
              underline={false}
              target='/account'
              renderLink={renderLink}>
              My Account
            </Link>
            { (isSubscriptionMember && childCount < 4) &&
              <React.Fragment>
                <Link
                  uppercase
                  underline={false}
                  target='/add-kid'
                  renderLink={renderLink}>
                  Add a Big Kid (2-14)
                </Link>
                <Link
                  uppercase
                  underline={false}
                  target='/mini/quiz'
                  renderLink={renderLink}>
                  Add a Baby (0-24M)
                </Link>
              </React.Fragment>
            }
            {isSubscriptionMember &&
              <Link
                uppercase
                underline={false}
                target='/invite'
                renderLink={renderLink}>
                Refer a Friend
              </Link>
            }
            <Link
              uppercase
              underline={false}
              onClick={signOut}>
              Log Out
            </Link>
          </nav>
        }
      </CSSTransitionGroup>
    )
  }
}

const SubMenu = styled(BaseSubMenu)`
  ${props => !props.inline ? position : ''}
  .subMenu {
    width: 225px;
    border: 1px solid ${props => props.theme.colors.rocketBlue};
    box-shadow: 0 2px 4px 0 rgba(0,0,0,.4);
    background-color: ${props => props.theme.colors.white};
      ${Link} {
        height: 50px;
        padding: 0 26px;
        display: flex;
        align-items: center;
        box-sizing: border-box;
        &:hover {
          background-color: ${props => props.theme.colors.rocketBlue};
          color: ${props => props.theme.colors.white};
          transition: all 0.25s ease-in-out;
        }
      }
    }

  .highlighted {
    background-color: ${props => props.theme.colors.rocketBlue};
    color: ${props => props.theme.colors.white};
  }

  .enter {
    z-index: 4;
    animation: ${slideDown} 0.25s;
    transform: translateY(-1px);
  }
  .leave {
    z-index: 4;
    transition: transform 0.25s ease-in;
    transform: translateY(-60px);
  }
`


SubMenu.propTypes = {
  open: PropTypes.bool,
  signOut: PropTypes.func,
  childCount: PropTypes.number,
  isSubscriptionMember: PropTypes.bool,
  pathname: PropTypes.string
}

SubMenu.defaultProps = {
  childCount: 0,
  isSubscriptionMember: false,
  signOut: () => alert('Signing Out')
}

/** @component */
export default SubMenu
