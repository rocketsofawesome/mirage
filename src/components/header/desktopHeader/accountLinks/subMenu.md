_Note: The `inline` prop is simply used in the style guide_

### Not Opened:
```js
  <SubMenu inline />
```

### Non Subscriber:
```js
  <SubMenu open inline />
```

### Subscriber with less than 4 children:
```js
  <SubMenu childCount={1} open inline showSubscriptionLinks />
```

### Subscriber with 4 children:
```js
  <SubMenu childCount={4} open inline showSubscriptionLinks />
```
