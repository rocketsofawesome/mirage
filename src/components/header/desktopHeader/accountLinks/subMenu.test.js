import React from 'react'
import 'jest-styled-components'

import { SubMenu, Link } from 'SRC'
import { BaseSubMenu } from './subMenu'

const { shallowWithTheme, mountWithTheme } = global

const loggedInNonSubscriber = {
  loggedIn: true
}

describe('(Styled) SubMenu', () => {
  const createSubMenu = (props) => {
    return shallowWithTheme(<SubMenu {...props} />)
  }

  test('matching the snapshot', () => {
    expect(createSubMenu())
    .toMatchSnapshot()
  })
})

describe('(Base Component) BaseSubMenu', () => {
  const createBaseSubMenu = (inProps) => {
    const props = {
      childCount: 0,
      isSubscriptionMember: false,
      ...inProps
    }
    return mountWithTheme(<BaseSubMenu {...props} />)
  }

  test('rendering no links when not open', () => {
    expect(createBaseSubMenu().contains(Link)).toBeFalsy()
  })

  test('rendering links when open', () => {
    expect(createBaseSubMenu({open: true}).contains(Link)).toBeTruthy()
  })

  test('not rendering subscription links if not subscriber', () => {
    const nonSubscriptionMenu = createBaseSubMenu({open: true})
    expect(nonSubscriptionMenu.contains('Manage Deliveries')).toBeFalsy()
    expect(nonSubscriptionMenu.contains('Style File')).toBeFalsy()
    expect(nonSubscriptionMenu.contains('Add a Big Kid (2-14)')).toBeFalsy()
    expect(nonSubscriptionMenu.contains('Add a Baby (0-24M)')).toBeFalsy()
  })

  test('rendering subscription links if a subscriber', () => {
    const nonSubscriptionMenu = createBaseSubMenu({
      open: true,
      isSubscriptionMember: true
    })
    expect(nonSubscriptionMenu.contains('Manage Deliveries')).toBeTruthy()
    expect(nonSubscriptionMenu.contains('Style File')).toBeTruthy()
    expect(nonSubscriptionMenu.contains('Add a Big Kid (2-14)')).toBeTruthy()
    expect(nonSubscriptionMenu.contains('Add a Baby (0-24M)')).toBeTruthy()
  })

  test('not rendering add a child link if a subscriber but with 4 kids', () => {
    const nonSubscriptionMenu = createBaseSubMenu({
      childCount: 4,
      open: true,
      isSubscriptionMember: true
    })
    expect(nonSubscriptionMenu.contains('Manage Deliveries')).toBeTruthy()
    expect(nonSubscriptionMenu.contains('Style File')).toBeTruthy()
    expect(nonSubscriptionMenu.contains('Add a Big Kid (2-14)')).toBeFalsy()
    expect(nonSubscriptionMenu.contains('Add a Baby (0-24M)')).toBeFalsy()
  })
})
