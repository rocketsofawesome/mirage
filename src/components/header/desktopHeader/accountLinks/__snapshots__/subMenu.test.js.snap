// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Styled) SubMenu matching the snapshot 1`] = `
.c0 {
  position: absolute;
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
  top: 100%;
  right: -1px;
  z-index: 10;
}

.c0 .subMenu {
  position: absolute;
  top: 100%;
  right: 0;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  height: auto;
}

.c0 .subMenu > * {
  width: 100%;
}

.c0 .subMenu {
  width: 225px;
  border: 1px solid #0073D1;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,.4);
  background-color: #FFFFFF;
}

.c0 .subMenu .c1 {
  height: 50px;
  padding: 0 26px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
}

.c0 .subMenu .c1:hover {
  background-color: #0073D1;
  color: #FFFFFF;
  -webkit-transition: all 0.25s ease-in-out;
  transition: all 0.25s ease-in-out;
}

.c0 .highlighted {
  background-color: #0073D1;
  color: #FFFFFF;
}

.c0 .enter {
  z-index: 4;
  -webkit-animation: cOMQan 0.25s;
  animation: cOMQan 0.25s;
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
}

.c0 .leave {
  z-index: 4;
  -webkit-transition: -webkit-transform 0.25s ease-in;
  -webkit-transition: transform 0.25s ease-in;
  transition: transform 0.25s ease-in;
  -webkit-transform: translateY(-60px);
  -ms-transform: translateY(-60px);
  transform: translateY(-60px);
}

<BaseSubMenu
  childCount={0}
  className="c0"
  isSubscriptionMember={false}
  signOut={[Function]}
/>
`;
